### PR TITLE
Bluetooth: controller: Make ULL/LLL split the default

### LINKED ISSUE #12681: BLE Split Link Layer
The "split link layer" is a version of the BLE controller which is split into:
* An upper, HW independent, LL (ULL)
* A lower, thin, LL (LLL), which depends on the particular HW being used

Currently this split LL is part of the master branch code, but is disabled ([`CONFIG_BT_LL_SW_SPLIT`](https://docs.zephyrproject.org/latest/reference/kconfig/CONFIG_BT_LL_SW_SPLIT.html?highlight=config_bt_ll_sw_split#cmdoption-arg-config-bt-ll-sw-split))

----------------------
This is the list of the missing features to reach parity with the old LL.

- [x] Build the split Link Layer in Zephyr CI (#12689)
- [x] Add support for Data Length Extensions (#13235): @cvinayak 
- [x] Add support for privacy (in platforms with HW address acceleration) (#14675)
- [x] Merge the new Link Layer into master (#12404)
- [x] Encapsulate the `req/ack` code blocks into helper functions ([link](https://github.com/zephyrproject-rtos/zephyr/pull/12404#pullrequestreview-194987023))
- [x] Update/improve controller documentation

-----------------------
Once parity is reached, this new link layer will be the default. The current target is for [release 1.15](https://github.com/zephyrproject-rtos/zephyr/projects/9)
Additional improvements are listed in #13519.

### PR DESCRIPTION
**Pull request by cvinayak**
Make the Upper and Lower Link Layer split architecture
implementation of the controller as the default when
building Zephyr Bluetooth Low Energy controller support
in applications.

Noticeable missing feature (porting) in comparison to old
architecture implementation is, Advanced scheduling of
connection events.

The missing features will subsequently be submitted
upstream.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

Fixes #12681
Depends on #16691

### CODE REVIEW

File: subsys/bluetooth/controller/ll_sw/ll.c [hunk 1]
```diff
@@ -175,7 +175,10 @@ int ll_init(struct k_sem *sem_rx)
 		return -ENOMEM;
 	}
 
+#if defined(CONFIG_BT_CTLR_FILTER)
```
Comment (jhedberg) [inline subsys/bluetooth/controller/ll_sw/ll.c]: Possible to use `IS_ENABLED()` here?

Comment (cvinayak) [reply subsys/bluetooth/controller/ll_sw/ll.c]: I included this change commit from #15780 just to get CI tests to work (as I added new conf files to testcases.yaml). I will remove the commit when #15780 gets merged first.

I kept the #15780 with the commit seperate in order to have backport label on that commit.

**Review by aescolar — DISMISSED** [review]
https://github.com/zephyrproject-rtos/zephyr/blob/master/tests/bluetooth/bsim_bt/bsim_test_app/prj.conf 
was meant to test the legacy controller,
and 
https://github.com/zephyrproject-rtos/zephyr/blob/master/tests/bluetooth/bsim_bt/bsim_test_app/prj_split.conf
the split one.
When changing the default, if we want to continue testing the legacy one, 
we would need to be set the 1st one to compile with BT_LL_SW.
(And whenever we are not interested in testing the legacy one anymore, we could just remove it)

**Review by carlescufi — DISMISSED** [review]
I will block this until we have confirmation that controller-based privacy is not a blocker for anyone, because this is a substantial feature that we are taking away from users.

**Comment by cvinayak** [comment]
@aescolar addressed the bsim test conf file in #15849

@aescolar @carlescufi moved commits out into new PR #15849, that are independent from switching the default controller implementation.

**Comment by tedd-an** [comment]
I have run PTS with #15849 + #15781 and here is the summary:

Summary:
The test result was same as the previous result which ran without this PRs and **didn't find any other failure**.
[pts_report.xlsx](https://github.com/zephyrproject-rtos/zephyr/files/3143884/pts_report.xlsx)

Some details about the result: The first run was based on the tip of the day + PRs and some GATT TCs failed and it seemed not related to this PR. After rebasing the baseline to the tip used on Wednesday test(and this PR), those failed tests were passed again. I will look into it separately.

**Comment by sixoctets** [comment]
@tedd-an Thank you for running the tests. Please upload the .config file of the zephyr build here (just to confirm it built the new link layer with CONFIG_BT_LL_SW_SPLIT=y).

**Comment by tedd-an** [comment]
attached .config.
[_config.txt](https://github.com/zephyrproject-rtos/zephyr/files/3148727/_config.txt)

**Comment by vkchettimada** [comment]
Thank you

File: subsys/bluetooth/controller/Kconfig [hunk 2]
```diff
@@ -16,12 +16,20 @@ if BT_CTLR
 
 choice BT_LL_CHOICE
 	prompt "Bluetooth Link Layer Selection"
-	default BT_LL_SW
 	help
 	  Select the Bluetooth Link Layer to compile.
 
+config BT_LL_SW_SPLIT
+	bool "Software-based BLE Link Layer (ULL/LLL split)"
```
Comment (joerchan) [inline subsys/bluetooth/controller/Kconfig]: Rename this perhaps?
Most people don't know what ULL/LLL split is, or care about that.

**Review by joerchan — CHANGES_REQUESTED** [review]
There are also a couple of bugfixes for the legacy controller, that should be ported to the new controller.

**Review by aescolar — APPROVED** [review]

File: subsys/bluetooth/controller/Kconfig (hunk 2 above)
Comment (cvinayak) [reply subsys/bluetooth/controller/Kconfig]: @joerchan suggestion?

Comment (joerchan) [reply subsys/bluetooth/controller/Kconfig]: I would suggest to rename this to BT_LL_SW and the old one to BT_LL_SW_LEGACY
Also remove the (ULL/LLL split) part of the description, and put (Legacy) in the description of the old one.

Comment (carlescufi) [reply subsys/bluetooth/controller/Kconfig]: I agree that is the better option too @cvinayak. That way the switch is seamless, although we do slightly change the meaning of a Kconfig option.

Comment (aescolar) [reply subsys/bluetooth/controller/Kconfig]: Just to join the bikeshedding: That would be a bit annoying, as we would need to change the places where today we use BT_LL_SW_SPLIT. Anyhow, chose whichever you guys prefer

Comment (cvinayak) [reply subsys/bluetooth/controller/Kconfig]: How about BT_LL_SW --> BT_LL_SW_LEGACY and keep the BT_LL_SW_SPLIT as is? in a subsequent PR, BT_LL_SW_SPLIT --> BT_LL_SW can be done.

File: doc/guides/bluetooth/bluetooth-arch.rst [hunk 3]
```diff
@@ -110,7 +110,7 @@ BLE-enabled builds that can be produced from the Zephyr project codebase:
   * :option:`CONFIG_BT_HCI` ``=y``
   * :option:`CONFIG_BT_HCI_RAW` ``=y``
   * :option:`CONFIG_BT_CTLR` ``=y``
-  * :option:`CONFIG_BT_LL_SW` ``=y`` (if using the open source Link Layer)
+  * :option:`CONFIG_BT_LL_SW_LEGACY` ``=y`` (if using the open source Link Layer)
```
Comment (cvinayak) [inline doc/guides/bluetooth/bluetooth-arch.rst]: fix this to use BT_LL_SW_SPLIT

File: doc/guides/bluetooth/bluetooth-arch.rst [hunk 4]
```diff
@@ -131,7 +131,7 @@ BLE-enabled builds that can be produced from the Zephyr project codebase:
   * :option:`CONFIG_BT` ``=y``
   * :option:`CONFIG_BT_HCI` ``=y``
   * :option:`CONFIG_BT_CTLR` ``=y``
-  * :option:`CONFIG_BT_LL_SW` ``=y`` (if using the open source Link Layer)
+  * :option:`CONFIG_BT_LL_SW_LEGACY` ``=y`` (if using the open source Link Layer)
```
Comment (cvinayak) [inline doc/guides/bluetooth/bluetooth-arch.rst]: fix this to use BT_LL_SW_SPLIT

File: drivers/counter/Kconfig.nrfx [hunk 5]
```diff
@@ -13,7 +13,7 @@ config COUNTER_NRF_RTC
 config COUNTER_TIMER0
 	bool "Enable Counter on TIMER0"
 	depends on HAS_HW_NRF_TIMER0
-	depends on !BT_LL_SW
+	depends on !BT_LL_SW_LEGACY
```
Comment (cvinayak) [inline drivers/counter/Kconfig.nrfx]: Add BT_LL_SW_SPLIT

File: drivers/counter/Kconfig.nrfx [hunk 6]
```diff
@@ -44,7 +44,7 @@ config COUNTER_TIMER4
 config COUNTER_RTC0
 	bool "Enable Counter on RTC0"
 	depends on HAS_HW_NRF_RTC0
-	depends on !BT_LL_SW
+	depends on !BT_LL_SW_LEGACY
```
Comment (cvinayak) [inline drivers/counter/Kconfig.nrfx]: Add BT_LL_SW_SPLIT

**Review by dbkinder — APPROVED** [review]
doc changes LGTM, other than your own comment to``use BT_LL_SW_SPLIT``

File: subsys/bluetooth/controller/Kconfig [hunk 7]
```diff
@@ -16,26 +16,27 @@ if BT_CTLR
 
 choice BT_LL_CHOICE
 	prompt "Bluetooth Link Layer Selection"
-	default BT_LL_SW
```
Comment (carlescufi) [inline subsys/bluetooth/controller/Kconfig]: Why did you remove this line?

Comment (cvinayak) [reply subsys/bluetooth/controller/Kconfig]: BT_LL_SW_SPLIT being the first item in the choice block, it is the de facto default

File: subsys/bluetooth/controller/Kconfig [hunk 8]
```diff
@@ -16,26 +16,27 @@ if BT_CTLR
...
 	help
 	  Select the Bluetooth Link Layer to compile.
 
-config BT_LL_SW_LEGACY
+config BT_LL_SW_SPLIT
```
Comment (carlescufi) [inline subsys/bluetooth/controller/Kconfig]: You need to rename this to `BT_LL_SW`, that's what we agreed with @joerchan right? no more "split" in the name

**Review by carlescufi — CHANGES_REQUESTED** [review]

File: subsys/bluetooth/controller/Kconfig (hunk 8 above)
Comment (cvinayak) [reply subsys/bluetooth/controller/Kconfig]: Using the same BT_LL_SW will make users’ config file be valid after switching to new LL but without them noticing the change. Users should know the change in design. I will keep the _SPLIT until legacy is deleted.

Comment (carlescufi) [reply subsys/bluetooth/controller/Kconfig]: That's really not the case if they are just using `CONFIG_BT` on a Nordic board, right? `BT_LL_SW_SPLIT` will be automatically enabled without them noticing

Comment (cvinayak) [reply subsys/bluetooth/controller/Kconfig]: Those who dont use explicit BT_LL_SW dont care if the macro is called BT_LL_SW_SPLIT or anything. We get them to use the new LL. Those who did use (for any reason) BT_LL_SW will now know that it is deprecated and take an informed decision to use BT_LL_SW_SPLIT.

**Review by carlescufi — APPROVED** [review]

**Review by joerchan — APPROVED** [review]
